src/Builder/Select/QueryExpressionVisitor.php

```diff
@@ -0,0 +1,199 @@
+                }
+
+                return sprintf('%s:%s', $field, $this->valueToString($value, ',', '"'));
+            case Comparison::LIKE:
+            case Comparison::MATCH:
+                if (\is_array($value)) {
+                    return sprintf('%s:(%s)', $field, $this->valueToString($value, ' OR ', '', false));
+                }
+
+                return sprintf('%s:%s', $field, $this->valueToString($value, ',', '', false));
+            case Comparison::NIN:
+                if (\is_array($value)) {
+                    return sprintf('-%s:(%s)', $field, $this->valueToString($value, ' OR ', '"'));
+                }
+
+                return sprintf('-%s:%s', $field, $this->valueToString($value, ',', '"'));
+            case Comparison::REGEXP:
+                if ('/' !== $value[0]) {
+                    $value = sprintf('/%s/', $value);
+                }
+
+                return sprintf('%s:%s', $field, $this->valueToString($value, ',', '', false));
+            case Comparison::EMPTY:
+                return sprintf('(*:* NOT %s:*)', $field);
+            default:
+                throw new RuntimeException(sprintf('Unknown comparison operator: %s', $expression->getOperator()));
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function walkValue(Value $value)
+    {
+        return $value->getValue();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \Solarium\Exception\RuntimeException
+     */
+    public function walkCompositeExpression(ExpressionInterface $expr)
+    {
+        $comparisons = [];
+
+        foreach ($expr->getComparisons() as $child) {
+            $comparisons[] = $this->dispatch($child);
+        }
+
+        switch ($expr->getType()) {
+            case CompositeComparison::TYPE_AND:
+                return implode(' AND ', $comparisons);
+            case CompositeComparison::TYPE_OR:
+                return implode(' OR ', $comparisons);
+            default:
+                throw new RuntimeException(sprintf('Unknown composite %s', $expr->getType()));
+        }
+    }
+
+    /**
+     * @param mixed  $value
+     * @param string $separator
+     * @param string $quote
+     * @param bool   $escape
+     *
+     * @return string
+     */
+    private function valueToString($value, string $separator = ',', string $quote = '', bool $escape = true): string
+    {
+        if (\is_array($value)) {
+            $ret = [];
+
+            foreach ($value as $v) {
+                $ret[] = $this->typedValueToString($v, $quote, $escape);
+            }
+
+            return implode($separator, $ret);
+        }
+
+        return $this->typedValueToString($value, $quote, $escape);
+    }
+
+    /**
+     * @param mixed  $value
+     * @param string $quote
+     * @param bool   $escape
+     *
+     * @return string
+     */
+    private function typedValueToString($value, string $quote = '', $escape = true): string
```

Comment:
One could theoretically wonder if `$escape` is nullable without a typehint.

```suggestion
    private function typedValueToString($value, string $quote = '', bool $escape = true): string
```